test/unit/test_keyring_raw_aes.py

```diff
@@ -0,0 +1,276 @@
+# Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+"""Unit tests for Raw AES keyring."""
+
+import os
+
+import mock
+import pytest
+from pytest_mock import mocker  # noqa pylint: disable=unused-import
+
+import aws_encryption_sdk.key_providers.raw
+import aws_encryption_sdk.keyring.raw_keyring
+from aws_encryption_sdk.identifiers import Algorithm, KeyringTraceFlag, WrappingAlgorithm
+from aws_encryption_sdk.internal.crypto.wrapping_keys import WrappingKey
+from aws_encryption_sdk.keyring.base import Keyring
+from aws_encryption_sdk.keyring.raw_keyring import GenerateKeyError, RawAESKeyring, _generate_data_key
+from aws_encryption_sdk.materials_managers import EncryptionMaterials
+from aws_encryption_sdk.structures import MasterKeyInfo
+
+from .unit_test_utils import (
+    _DATA_KEY,
+    _ENCRYPTED_DATA_KEY_AES,
+    _ENCRYPTED_DATA_KEY_NOT_IN_KEYRING,
+    _ENCRYPTION_CONTEXT,
+    _KEY_ID,
+    _PROVIDER_ID,
+    _SIGNING_KEY,
+    _WRAPPING_KEY,
+    get_decryption_materials_with_data_encryption_key,
+    get_decryption_materials_without_data_encryption_key,
+    get_encryption_materials_with_data_encryption_key,
+    get_encryption_materials_without_data_encryption_key,
+)
+
+pytestmark = [pytest.mark.unit, pytest.mark.local]
+
+
+@pytest.fixture
+def raw_aes_keyring():
+    return RawAESKeyring(
+        key_namespace=_PROVIDER_ID,
+        key_name=_KEY_ID,
+        wrapping_algorithm=WrappingAlgorithm.AES_256_GCM_IV12_TAG16_NO_PADDING,
+        wrapping_key=_WRAPPING_KEY,
+    )
+
+
+@pytest.fixture
+def patch_generate_data_key(mocker):
+    mocker.patch.object(aws_encryption_sdk.keyring.raw_keyring, "_generate_data_key")
+    return aws_encryption_sdk.keyring.raw_keyring._generate_data_key
+
+
+@pytest.fixture
+def patch_decrypt_on_wrapping_key(mocker):
+    mocker.patch.object(WrappingKey, "decrypt")
+    return WrappingKey.decrypt
+
+
+@pytest.fixture
+def patch_os_urandom(mocker):
+    mocker.patch.object(os, "urandom")
+    return os.urandom
+
+
+def test_parent():
+    assert issubclass(RawAESKeyring, Keyring)
+
+
+def test_valid_parameters(raw_aes_keyring):
+    test = raw_aes_keyring
+    assert test.key_name == _KEY_ID
+    assert test.key_namespace == _PROVIDER_ID
+    assert test._wrapping_algorithm == WrappingAlgorithm.AES_256_GCM_IV12_TAG16_NO_PADDING
+    assert test._wrapping_key == _WRAPPING_KEY
+
+
+@pytest.mark.parametrize(
+    "key_namespace, key_name, wrapping_algorithm, wrapping_key",
+    (
```

Comment:
The test cases I'd like to see here are:

- [ ] for each required parameter, that parameter's value is `None` but all other parameters are valid
- [ ] for each parameter with a validator, that parameter's value is invalid but all other parameters are valid (for required parameters there can be a lot of overlap between this and the previous where the same test case can address both)

For example, if I had a class that accepted parameters A, B, and C, of which A and B are required `int`s and C is an optional string, I would want:

```python
# A, B, C
None, 42, "petunias"
42, None, "petunias"
42, 42, 42
```